workflows/pipe-common/pipeline/hpc/gridengine.py

```diff
@@ -484,3 +546,226 @@ def validate(self, jobs):
                     continue
             valid_jobs.append(job)
         return valid_jobs, invalid_jobs
+
+
+class SlurmGridEngine(GridEngine):
+
+    _KILL_JOBS = "scancel %s"
+    _FORCE_KILL_JOBS = "scancel -f %s"
+    _SHOW_EXECUTION_HOST = "scontrol -o show node %s"
+    _SCONTROL_UPDATE_NODE_STATE = "scontrol update State=%s Reason='CP_CAP_GE: Autoscale lifecycle event.' NodeName=%s"
+    _SCONTROL_DELETE_NODE = "scontrol delete nodename=%s"
+    _SCONTROL_PARSE_HOSTLIST = "scontrol show hostnames %s"
+
+    _NODE_BAD_STATES = ["DOWN", "DRAINING", "DRAIN", "DRAINED", "FAIL", "FAILING",  "INVAL"]
+
+    _SCONTROL_DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S'
+    _GET_JOBS = "scontrol -o show job"
+
+    def __init__(self, cmd_executor):
+        self.cmd_executor = cmd_executor
+
+    def get_jobs(self):
+        try:
+            output = self.cmd_executor.execute(SlurmGridEngine._GET_JOBS)
+        except ExecutionError:
+            Logger.warn('Slurm jobs listing has failed.')
+            return []
+        return self._parse_jobs(output)
+
+    def disable_host(self, host):
+        self.cmd_executor.execute(SlurmGridEngine._SCONTROL_UPDATE_NODE_STATE % ("DRAIN", host))
+
+    def enable_host(self, host):
+        host_state = self._get_host_state(host)
+        if "DRAIN" in host_state:
+            self.cmd_executor.execute(SlurmGridEngine._SCONTROL_UPDATE_NODE_STATE % ("UNDRAIN", host))
+        else:
+            # NO-OP for all node states except DRAIN
+            pass
+
+    def get_pe_allocation_rule(self, pe):
+        raise AllocationRuleParsingError("Slurm doesn't have PE preference.")
+
+    def delete_host(self, host, skip_on_failure=False):
+        _perform_command(
+            action=lambda: self.disable_host(host),
+            msg='Disabling GE host execution daemon.',
+            error_msg='Disabling GE host execution daemon has failed.',
+            skip_on_failure=skip_on_failure
+        )
```

Comment:
No need to disable host here because `self.disable_host` should be called beforehand by `GridEngineScaleDownHandler`.